src/phoenix/server/api/routers/trace_handler.py

```diff
@@ -54,7 +60,122 @@ async def post(self, request: Request) -> Response:
-                for span in scope_span.spans:
-                    self.traces.put(decode(span), project_name=project_name)
+                for otlp_span in scope_span.spans:
+                    span = decode(otlp_span)
+                    async with self.db() as session:
+                        await _insert_span(session, span, project_name)
+                    self.traces.put(span, project_name=project_name)
                     await asyncio.sleep(0)
         return Response()
+
+
+async def _insert_span(session: AsyncSession, span: Span, project_name: str) -> None:
+    if not (
+        project_rowid := await session.scalar(
+            text("SELECT rowid FROM projects WHERE name = :name;"),
+            {"name": project_name},
+        )
+    ):
+        project_rowid = await session.scalar(
+            text("INSERT INTO projects(name) VALUES(:name) RETURNING rowid;"),
+            {"name": project_name},
+        )
+    if (
+        trace_rowid := await session.scalar(
+            text(
+                """
+                INSERT INTO traces(trace_id, project_rowid, session_id, start_time, end_time)
+                VALUES(:trace_id, :project_rowid, :session_id, :start_time, :end_time)
+                ON CONFLICT DO UPDATE SET
+                start_time = CASE WHEN excluded.start_time < start_time THEN excluded.start_time ELSE start_time END,
+                end_time = CASE WHEN end_time < excluded.end_time THEN excluded.end_time ELSE end_time END
+                WHERE excluded.start_time < start_time OR end_time < excluded.end_time
+                RETURNING rowid;
+                """  # noqa E501
+            ),
+            {
+                "trace_id": span.context.trace_id,
+                "project_rowid": project_rowid,
+                "session_id": None,
+                "start_time": span.start_time,
+                "end_time": span.end_time,
+            },
+        )
+    ) is None:
+        trace_rowid = await session.scalar(
+            text("SELECT rowid from traces where trace_id = :trace_id"),
+            {"trace_id": span.context.trace_id},
+        )
+    cumulative_error_count = int(span.status_code is SpanStatusCode.ERROR)
+    cumulative_llm_token_count_prompt = cast(
+        int, span.attributes.get(SpanAttributes.LLM_TOKEN_COUNT_PROMPT, 0)
+    )
+    cumulative_llm_token_count_completion = cast(
+        int, span.attributes.get(SpanAttributes.LLM_TOKEN_COUNT_COMPLETION, 0)
+    )
+    if accumulation := (
+        await session.execute(
+            text(
+                """
+                SELECT
+                sum(cumulative_error_count),
+                sum(cumulative_llm_token_count_prompt),
+                sum(cumulative_llm_token_count_completion)
+                FROM spans
+                WHERE parent_span_id = :parent_span_id
+                """
+            ),  # noqa E501
+            {"parent_span_id": span.context.span_id},
+        )
+    ).first():
```

Comment:
this also feels a tad in-efficient - where a read needs to be performed before you insert. If anything you should be able to execute all precursor work on a single async gather call so that the db calls are parallelized?

Reply:
This is probably fine assuming that the child spans are still in the buffer pool. On the other hand, parallelization doesn't really help sqlite because it locks the whole database